### PR TITLE
alternator: support "deflate" encoding in request compression

### PR DESCRIPTION
**Pull request by nyh**
Currently Alternator supports compressed requests in the gzip format with "Content-Encoding: gzip". We did not support any other compression formats.

It turns out that DynamoDB also supports the "deflate" encoding. Because the "deflate" format is a small variant of gzip and also supported by the same zlib library we already use, it is very easy to add support for it as well. So this patch adds it.

Beyond compatibility with DynamoDB, another benefit of this patch is symmetry with our response compression support (PR #27454), where we supported both gzip and deflate compression of responses - so we should support the same for requests.

This patch also adds a test for Content-Encoding: deflate, which passes on DynamoDB (proving that "deflate" is indeed supported there). On Alternator the new test failed before this patch and passes with this patch.

Refs #27243 (which asks to support more compression formats).

No need to backport - this is a small, esoteric, new feature, and comes on top of the request compression feature which itself did not get backported.

### CODE REVIEW

File: alternator/server.cc [hunk 1]
```diff
@@ -586,15 +586,17 @@ class safe_gzip_zstream {
...
+ungzip(chunked_content&& compressed_body, size_t length_limit, bool gzip = true) {
     chunked_content ret;
     // output_buf can be any size - when uncompressing input_buf, it doesn't
     // need to fit in a single output_buf, we'll use multiple output_buf for
     // a single input_buf if needed.
     constexpr size_t OUTPUT_BUF_SIZE = 4096;
     temporary_buffer<char> output_buf;
-    safe_gzip_zstream strm;
+    safe_gzip_zstream strm(gzip);
     bool complete_stream = false; // empty input is not a valid gzip
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/server.cc]: Good point, I also need to add a test for empty deflate, to make sure it is really invalid also in DynamoDB.

File: alternator/server.cc [hunk 2]
```diff
@@ -586,15 +586,17 @@ class safe_gzip_zstream {
 // ungzip() takes a chunked_content with a gzip-compressed request body,
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/server.cc]: I'll rewrite the comment. Perhaps I should have renamed the function (which is still called ungzip()), but I would rather not do this right now. The name "gzip" isn't that misleading anyway - the differences between gzip and zlib (deflate) are minor anyway.

File: test/alternator/test_compressed_request.py [hunk 3]
```diff
@@ -275,3 +276,21 @@ def test_gzip_request_empty(dynamodb):
...
+    v = random_string()
+    payload = '{"TableName": "' + test_table.name + '", "Item": {"p": {"S": "' + p + '"}, "c": {"S": "x"}, "v": {"S": "' + v + '"}}}'
+    payload = zlib.compress(payload.encode('utf-8'))
+    req = get_signed_request(dynamodb, 'PutItem', payload)
+    headers = dict(req.headers)
+    headers.update({'Content-Encoding': 'deflate'})
+    r = requests.post(req.url, headers=headers, data=req.body, verify=False)
+    assert r.status_code == 200, f'Request failed: {r.content}'
+    got = test_table.get_item(Key={'p': p, 'c': 'x'}, ConsistentRead=True)['Item']
+    assert got == {'p': p, 'c': 'x', 'v': v}
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/alternator/test_compressed_request.py:296]: I wanted to avoid doing all of this, considering that the two implementations are almost identical anyway. But on second thought, it's a good idea. I'll add more tests. For some of them the result isn't even obvious (is it really allowed to concatenate two deflate outputs, not just two gzips), so it's good to test.

**Comment by nyh** [comment]
Pushed a new version implementing @copilot's suggestions: Improved several comments, and added more tests.
